Adding common transforms for Swift in iOS

tchype opened the pull request:
*Issue #, if available:* #161 

*Description of changes:*
Adding ability to use `ios-swift` and `ios-swift-separate` transform groups.  Examples:

*ios-swift*: _all in one dictionary class file_

```swift
//
// StyleDictionary.swift
//
// Do not edit directly
// Generated on Tue, 19 Mar 2019 18:31:55 GMT
//

import UIKit

public class StyleDictionary {
    static let colorBaseGrayDark = UIColor(red: 0.07, green: 0.07, blue: 0.07, alpha:1)
    static let colorBaseGrayLight = UIColor(red: 0.80, green: 0.80, blue: 0.80, alpha:1)
    static let colorBaseGrayMedium = UIColor(red: 0.60, green: 0.60, blue: 0.60, alpha:1)
    static let colorBaseGreen = UIColor(red: 0.00, green: 1.00, blue: 0.00, alpha:1)
    static let colorBaseRed = UIColor(red: 1.00, green: 0.00, blue: 0.00, alpha:1)
    static let colorFontBase = UIColor(red: 1.00, green: 0.00, blue: 0.00, alpha:1)
    static let colorFontSecondary = UIColor(red: 0.00, green: 1.00, blue: 0.00, alpha:1)
    static let colorFontTertiary = UIColor(red: 0.80, green: 0.80, blue: 0.80, alpha:1)
    static let sizeFontBase = CGFloat(16.00)
    static let sizeFontLarge = CGFloat(32.00)
    static let sizeFontMedium = CGFloat(16.00)
    static let sizeFontSmall = CGFloat(12.00)
}
```

*ios-swift-separate*: _allow enum-looking use so each category is separated into different files_

```swift
//
// StyleDictionaryColor.swift
//
// Do not edit directly
// Generated on Sun, 10 Mar 2019 23:02:19 GMT
//

import UIKit

public enum StyleDictionaryColor {
    static let baseGrayLight = UIColor(red: 0.80, green: 0.80, blue: 0.80, alpha:1)
    static let baseGrayMedium = UIColor(red: 0.60, green: 0.60, blue: 0.60, alpha:1)
    static let baseGrayDark = UIColor(red: 0.07, green: 0.07, blue: 0.07, alpha:1)
    static let baseRed = UIColor(red: 1.00, green: 0.00, blue: 0.00, alpha:1)
    static let baseGreen = UIColor(red: 0.00, green: 1.00, blue: 0.00, alpha:1)
    static let fontBase = UIColor(red: 1.00, green: 0.00, blue: 0.00, alpha:1)
    static let fontSecondary = UIColor(red: 0.00, green: 1.00, blue: 0.00, alpha:1)
    static let fontTertiary = UIColor(red: 0.80, green: 0.80, blue: 0.80, alpha:1)
}
```
```swift
//
// StyleDictionarySize.swift
//
// Do not edit directly
// Generated on Sun, 10 Mar 2019 23:02:19 GMT
//

import UIKit

public enum StyleDictionarySize {
    static let fontSmall: CGFloat = 12.00
    static let fontMedium: CGFloat = 16.00
    static let fontLarge: CGFloat = 32.00
    static let fontBase: CGFloat = 16.00
}
```

For `ios-swift-separate` (as shown above), it creates public enums but adds static members to the enum class without any `case` statements.  If we were to type the enum and then set values, every time someone wants to use them they would have to do something like:

```swift
let size = StyleDictionarySize.fontBase.rawValue
```
By doing it the way we do in the proposed changes, we allow for: 

```swift
let size = StyleDictionary.fontBase
```

Also added the following transforms:
 - `name/ti/camel`: supports camel-case without the leading category (Swift encourages succinct names); also added supporting tests.
 - `size/swift/remToCGFloat`: like the already-in-place `color/UIColorSwift`, this adds support for `CGFloat`, which is used in UIKit for decimal values; swift will view a literal number with decimal point values as a `Double` without a class initializer.
- `content/swift/literal`: the `content/objC/literal` prepends the `@`; swift doesn't use this for literals.
- `font/swift/literal`: the `font/objC/literal` prepends the `@`; swift doesn't use this for literals.
- `asset/swift/literal`: the `asset/objC/literal` prepends the `@`; swift doesn't use this for literals.

Added the `ios-swift` and `ios-swift-separate` transform groups using the new transforms as appropriate.

Added support for two specific templates:

- `ios-swift/class.swift`: leverages Swift's type inference to allow any literal that is supported (e.g., String, UIColor, CGFloat) to be added as a property to a class.
- `ios-swift/enum.swift`: leverages Swift's type inference to allow any literal that is supported (e.g., String, UIColor, CGFloat) but puts it into an enum instead of a class.

Updated the `basic` example to show size and color to match the Objective-C version by leveraging `ios-swift-separate` and also added a single `ios-swift` output to show how they can all be used in one `StyleDictionary` class instead.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

dbanksdesign commented:
This is awesome! I will take a look this week, but thank you so much for doing this!

davixyz commented:
👍

tchype commented on lib/common/transforms.js:
```diff
@@ -704,5 +786,25 @@ module.exports = {
+      return prop.attributes.category === 'asset';
+    },
+    transformer: function(prop) {
+      return '"' + prop.value + '"';
```
Yup.  Plus, I'll add an isAsset test function since that's copy/pasted everywhere as well.

tchype commented on lib/common/transforms.js:
````diff
@@ -623,6 +684,27 @@ module.exports = {
+   *
+   * ```swift
+   * // Matches: prop.attributes.category === 'font'
+   * // Returns: @"string"
````
Yup, copy/paste error.

tchype commented on lib/common/templates/ios-swift/static.swift.template:
```diff
@@ -0,0 +1,35 @@
+//
+
+<%  var props = _.filter(allProperties, this.filter);
+    var type = this.type; %>
```
Nah, copy/paste from other template I used as a starting point.  I thought it was used by other embedded logic in the file, but nope 👍

tchype commented on lib/common/transformGroups.js:
```diff
@@ -168,6 +168,28 @@ module.exports = {
+   * Transforms:
+   *
+   * [attribute/cti](transforms.md#attributecti)
+   * [name/cti/pascal](transforms.md#namectipascal)
```
Updated this and also added sections to the transforms.md file to match.

tchype commented:
Review comments addressed, added `isAsset` along with helper methods for wrapping prop values with characters like single and double quotes. Updated `transforms.md` with the new transforms and fixed a couple copy/paste errors.

chazzmoney commented:
Hey, this is GREAT! We have wanted to add Swift support for awhile (I'm pretty amateur w/ Swift), so thank you for doing this.

I do have one concern; the setup appears to require the user to output all sizes in one file and all other properties in another. It requires this decision making to happen in the config. Instead, it would be great if there was only a single Swift format and the properties themselves were used to determine when/how to add in type information.

Maybe we create a new transform to add an attribute containing the token type (tokenType? classType?). Then when the formatter is run, it can use that information to output the type info correctly.

Does that make sense? Or am I misinterpreting something about Swift?

tchype commented:
Honestly, that's a good point.  I had considered that, with the way "autocomplete/intellisense" works in XCode, that having to type `StyleDictionary.color` or `StyleDictionary.size` is a lot to type to still have a ton of options.  That being said, before merging this, let me play with some ideas that can make it more flexible and potentially even allow folks to choose one or the other.

It's the transform groups where we essentially "decide" the conventions for the language, so I'll play with it and get back to you.

tchype commented:
I've added the default `ios-swift` transform group that can support just adding everything that meets the filter into one class.  I also added `ios-swift-separate` to support having separate classes and/or enums by filter but that doesn't have the category name in the name of the property.  (see the pull request description for the two examples).  Ready for you to accept if you are OK with it, @dbanksdesign !

dbanksdesign commented:
My apologies for the lateness, I am currently in Europe right now. I will dig into the code tonight (in about 12 hours), but reading over your comment and description it sounds good! Thank you again for all your work on this.

dbanksdesign commented on lib/common/templates/ios-swift/enum.swift.template:
```diff
@@ -0,0 +1,39 @@
+  // Filter to only those props wanted based on the filter, then sort
+  // them by category so we keep like props together, then by name
+  // so they are easier to find alphabetically.
+  var filteredProps = _.filter(allProperties, this.filter); 
```
I'm 99% sure you can remove this, I believe when we properly added the filter functionality we actually run the filters before the format so the format code itself doesn't have to worry about implementing filters. [Take a look here](https://github.com/amzn/style-dictionary/blob/master/lib/buildFile.js#L45)

dbanksdesign commented on lib/common/transforms.js:
```diff
@@ -578,7 +628,7 @@ module.exports = {
     },
     transformer: function(prop) {
-      return '\'' + prop.value + '\'';
+      return  wrapValueWith('\'', prop);
```
Love this addition!

tchype commented:
Thanks!  Will make the updates and get back to you.  Thanks again!


On Sat, Mar 23, 2019 at 2:07 PM Danny Banks <notifications@github.com>
wrote:

> *@dbanksdesign* commented on this pull request.
>
> 2 super minor things, after that I'm good to go. Thank you again for your
> work on this!
> ------------------------------
>
> In examples/basic/config.json
> <https://github.com/amzn/style-dictionary/pull/255#discussion_r268409267>:
>
> > @@ -64,6 +64,42 @@
>            }
>          }
>        }]
> +    },
> +    "ios-swift": {
> +      "transformGroup": "ios-swift",
> +      "buildPath": "build/ios-swift/",
> +      "files": [{
> +        "destination": "StyleDictionary.swift",
> +        "format": "ios-swift/class.swift",
> +        "className": "StyleDictionary",
> +        "type": "StyleDictionaryColorName",
>
> It still doesn't look like 'type' here is being used, I can't see it in
> the format code. Is this meant to do something or could we remove it?
> ------------------------------
>
> In lib/common/templates/ios-swift/enum.swift.template
> <https://github.com/amzn/style-dictionary/pull/255#discussion_r268409475>:
>
> > +// <%= this.destination %>
> +//
> +<%
> +    // for backward compatibility we need to have the user explicitly hide it
> +    var showFileHeader = (this.options && this.options.hasOwnProperty('showFileHeader')) ? this.options.showFileHeader : true;
> +    if(showFileHeader) {
> +        print("// Do not edit directly\n");
> +        print("// Generated on " + new Date().toUTCString());
> +    }
> +%>
> +//
> +<%
> +  // Filter to only those props wanted based on the filter, then sort
> +  // them by category so we keep like props together, then by name
> +  // so they are easier to find alphabetically.
> +  var filteredProps = _.filter(allProperties, this.filter);
>
> I'm 99% sure you can remove this, I believe when we properly added the
> filter functionality we actually run the filters before the format so the
> format code itself doesn't have to worry about implementing filters. Take
> a look here
> <https://github.com/amzn/style-dictionary/blob/master/lib/buildFile.js#L45>
> ------------------------------
>
> In lib/common/transforms.js
> <https://github.com/amzn/style-dictionary/pull/255#discussion_r268409496>:
>
> > @@ -578,7 +628,7 @@ module.exports = {
>        return prop.attributes.category === 'content';
>      },
>      transformer: function(prop) {
> -      return '\'' + prop.value + '\'';
> +      return  wrapValueWith('\'', prop);
>
> Love this addition!
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/amzn/style-dictionary/pull/255#pullrequestreview-218055853>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAObpWV4Wbc8WSPcKQD-A5FJgSfZSPD-ks5vZpeogaJpZM4bnl85>
> .
>


-- 

*when you don’t create things, you become defined by your tastes rather
than ability. your tastes only narrow & exclude people. so create.*
- _why

chazzmoney commented:
@tchype Thank again for all your hard work on this. We would love to get this functionality merged in.

tchype commented:
Sorry -- thought I finished this.  Will finish today @chazzmoney !

tchype commented:
Ready for review/merge @chazzmoney @dbanksdesign .  Thanks!

chazzmoney commented:
LGTM! :shipit:
